solarforecastarbiter/reference_forecasts/persistence.py

```diff
@@ -0,0 +1,272 @@
+# models_persistence.py?
+"""
+Functions for persistence forecasts.
+"""
+from functools import partial
+
+import numpy as np
+import pandas as pd
+
+from solarforecastarbiter import datamodel, pvmodel
+from solarforecastarbiter.io.utils import load_data  # does not exist
+from solarforecastarbiter.reference_forecasts import forecast
+
+
+def persistence(observation, window, data_start, data_end,
+                forecast_start, forecast_end, interval_length,
+                load_data=load_data):
+    """
+    Make a persistence forecast for the observation.
+
+    Be careful with start times, end times, window, and interval length.
+    If persistence of a scalar quantity is desired, data_end -
+    data_start must be less than (interval label = instant) or less than
+    or equal (interval label = beginning or ending) to window. If
+    persistence of multiple values is desired, data_end - data_start
+    must equal to forecast_end - forecast_start.
+
+    Parameters
+    ----------
+    observation : datamodel.Observation
+    window : pd.Timedelta
+        Time period over which to calculate the persistence quantity
+    data_start : pd.Timestamp
+        Data start
+    data_end : pd.Timestamp
+        Data end
+    forecast_start : pd.Timestamp
+        Forecast start
+    forecast_end : pd.Timestamp
+        Forecast end
+    interval_length : pd.Timedelta
+        Forecast interval length
+    load_data : function
+        A function that loads the observation data.
+
+    Returns
+    -------
+    forecast : pd.Series
+        The persistence forecast. The forecast interval label equal to
+        the observation interval label.
+
+    Raises
+    ------
+    ValueError
+        If window > data_end - data_start and data_end - data_start !=
+        forecast_end - forecast_start.
+
+    ValueError
+        If calculated forecast length is inconsistent with desired
+        forecast length.
+    """
+    # consistency checks for labels, data duration, and forecast duration
+    closed = _check_interval_closure(
+        observation.interval_label, window, data_start, data_end,
+        forecast_start, forecast_end, interval_length)
+    # assumes data (possibly nan) exists from start to end
+    obs = load_data(observation, data_start, data_end)
+    # only support mean for now
+    persistence_quantity = obs.resample(window, closed=closed).mean()
```

Comment:
Let me make sure I follow you. Are you saying that this behavior is wrong?

```python
In [15]: data_start, data_end = '20190412T1210', '20190412T1225'

In [16]: window = '15min'

In [17]: closed = 'left'

In [18]: s = pd.Series([1, 2, 3], index=pd.date_range(start=data_start, end=data_end, freq='5min', closed=closed))

In [19]: s
Out[19]:
2019-04-12 12:10:00    1
2019-04-12 12:15:00    2
2019-04-12 12:20:00    3
Freq: 5T, dtype: int64

In [20]: s.resample(window, closed='left').mean()
Out[20]:
2019-04-12 12:00:00    1.0
2019-04-12 12:15:00    2.5
Freq: 15T, dtype: float64
```

and that instead you'd want to see

```python
In [20]: s.resample(window, closed='left').mean()
Out[20]:
2019-04-12 12:10:00    2.0
Freq: 15T, dtype: float64
```

Reply:
If we are saying that we'll average the dta between start=12:10 and end=12:25 at 15m intervals, then I would not expect the first value to be timestamped at 12:00:00.  Using left labeling, I'd expect the first point to be timestamped 12:10 with a value of 1.33 (1 for 10 minutes, 2 for 5 minutes).

Reply:
Ok, I agree.

> We haven't discussed weighting if the window is not evenly divided by the data interval.

I'd be ok with not supporting situations in which weighting would be necessary. In the current implementation it would raise a `ValueError` (I think).

Reply:
I think this problem will be somewhat if not entirely mitigated if we pursue the additional functions as described in the [comments](https://github.com/SolarArbiter/solarforecastarbiter-core/pull/59#discussion_r275068122) below.

For the `persistence_scalar` function, we know we want to reduce the data to the mean, so we can simply call `obs.mean()`. For the `persistence_interval` function, I think we can enforce start and end times that don't cause problems. I don't think the `*_index` functions will be any different.